File: Python3/1138.py
__________________________________________________________________________________________________
sample 24 ms submission
class Solution:
    def alphabetBoardPath(self, target: str) -> str:
        code = [ord(t) - ord('a') for t in target]
        code = [(c // 5, c % 5) for c in code]
        res = ""
        code = [(0,0)] + code
        
        def path(step1, step2, num, res):
            if num > 0:
                res += step1 * num 
            elif num < 0:
                res += step2 * (-num)
            return res
        
        for i in range(1, len(code)):
            v = code[i][0] - code[i-1][0]
            h = code[i][1] - code[i-1][1]
            if code[i] == (5,0):
                res = path('R', 'L', h, res)
                res = path('D', 'U', v, res)
            else:
                res = path('D', 'U', v, res)
                res = path('R', 'L', h, res)
            res += '!'
        return res
__________________________________________________________________________________________________
sample 28 ms submission
class Solution:
    def alphabetBoardPath(self, target: str) -> str:
        s,i,j,n='',0,0,ord('a')
        for c in target:
            t=ord(c)-n
            ii,jj=t//5,t%5
            if ii>i:
                if jj>j: s+='R'*(jj-j)
                else: s+='L'*(j-jj)
                s+='D'*(ii-i)
            else:
                s+='U'*(i-ii)
                if jj>j: s+='R'*(jj-j)
                else: s+='L'*(j-jj)
            s+='!'
            i,j=ii,jj
        return s
__________________________________________________________________________________________________
sample 32 ms submission
class Solution:
    def alphabetBoardPath(self, target: str) -> str:
        board = ["abcde", "fghij", "klmno", "pqrst", "uvwxy", "z"]
        adjust = ord('a')
        words_per_set = 5
        current_r=0
        current_c=0
        result=""
        
        def is_exits(r,c):
            try:
                board[r][c]
                return True
            except Exception:
                return False
            
        for t_c in target:
            adj_ord_c=ord(t_c)-adjust
            target_r = adj_ord_c // words_per_set
            target_c = adj_ord_c % words_per_set
            
            while (target_r != current_r) or (target_c != current_c):
                if current_r > target_r and is_exits(current_r-1,current_c):
                    result+="U"
                    current_r -= 1
                    continue
                if current_r < target_r and is_exits(current_r+1,current_c):
                    result+="D"
                    current_r += 1
                    continue
                if current_c > target_c and is_exits(current_r,current_c-1):
                    result+="L"
                    current_c -= 1
                    continue
                if current_c < target_c and is_exits(current_r,current_c+1):
                    result+="R"
                    current_c += 1
                    continue
            
            result += "!"
            
        
        return result

File: C++/1138.cpp
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
public:
    
    string alphabetBoardPath(string target) {
        string t = "a" + target;
        string answer;
        for (int i = 0; i < t.size() - 1; ++i) {
            answer += solvetwo(t[i], t[i + 1]);
        }
        return answer;
    }

private:
    unordered_map<char, vector<int>> char_coords = {{'a',{0,0}}, {'b',{0,1}}, {'c',{0,2}}, {'d',{0,3}}, {'e',{0,4}}, {'f',{1,0}}, {'g',{1,1}}, {'h',{1,2}}, {'i',{1,3}}, {'j',{1,4}}, {'k',{2,0}}, {'l',{2,1}}, {'m',{2,2}}, {'n',{2,3}}, {'o',{2,4}}, {'p',{3,0}}, {'q',{3,1}}, {'r',{3,2}}, {'s',{3,3}}, {'t',{3,4}}, {'u',{4,0}}, {'v',{4,1}}, {'w',{4,2}}, {'x',{4,3}}, {'y',{4,4}}, {'z',{5,0}}};
    
    unordered_map<string, string> answer_map;
    
    string solvetwo(char from, char to) {
        if (from == to) {
            return "!";
        }        
        vector<int>& x_coords = char_coords[from];
        vector<int>& y_coords = char_coords[to];
        vector<char> answer;
        if (from != 'z') {
            int d = y_coords[1] - x_coords[1];
            if (d > 0) {
                for (int i = 0; i < d; ++i) {
                    answer.push_back('R');
                }
            } else if (d < 0) {
                d = -1*d;
                for (int i = 0; i < d; ++i) {
                    answer.push_back('L');
                }
            }
            d = y_coords[0] - x_coords[0];
            if (d > 0) {
                for (int i = 0; i < d; ++i) {
                    answer.push_back('D');
                }
            } else if (d < 0) {
                d = -1*d;
                for (int i = 0; i < d; ++i) {
                    answer.push_back('U');
                }
            } 
            answer.push_back('!');
        } else {
            int d = y_coords[0] - x_coords[0];
            if (d > 0) {
                for (int i = 0; i < d; ++i) {
                    answer.push_back('D');
                }
            } else if (d < 0) {
                d = -1*d;
                for (int i = 0; i < d; ++i) {
                    answer.push_back('U');
                }
            }
            d = y_coords[1] - x_coords[1];
            if (d > 0) {
                for (int i = 0; i < d; ++i) {
                    answer.push_back('R');
                }
            } else if (d < 0) {
                d = -1*d;
                for (int i = 0; i < d; ++i) {
                    answer.push_back('L');
                }
            }             
            answer.push_back('!');
        }
        
        return string(answer.begin(), answer.end());
    }
};
__________________________________________________________________________________________________
class Solution {
public:
    string alphabetBoardPath(string target, int x = 0, int y = 0) {
  string res;
  for (auto ch : target) {
    int x1 = (ch - 'a') % 5, y1 = (ch - 'a') / 5;
    res += string(max(0, y - y1), 'U') + string(max(0, x1 - x), 'R') +
      string(max(0, x - x1), 'L') + string(max(0, y1 - y), 'D') + "!";
    x = x1, y = y1;
  }
  return res;
}
};
__________________________________________________________________________________________________
